File: src/main/java/io/gitlab/intended/storagenetworks/registry/BlockEntityTypeRegistry.java
package io.gitlab.intended.storagenetworks.registry;

import io.gitlab.intended.storagenetworks.block.ModBlock;
import io.gitlab.intended.storagenetworks.block.entity.*;
import net.minecraft.block.Block;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.util.registry.Registry;

import java.util.function.Supplier;

public final class BlockEntityTypeRegistry {

    public static final BlockEntityType<MasterBlockEntity> MASTER = create(MasterBlockEntity::new, BlockRegistry.MASTER);
    public static final BlockEntityType<CraftingTerminalBlockEntity> CRAFTING_TERMINAL = create(CraftingTerminalBlockEntity::new, BlockRegistry.CRAFTING_TERMINAL);
    public static final BlockEntityType<ProcessingTerminalBlockEntity> PROCESSING_TERMINAL = create(ProcessingTerminalBlockEntity::new, BlockRegistry.PROCESSING_TERMINAL);

    public static final BlockEntityType<CableBlockEntity> CABLE = create(CableBlockEntity::new, BlockRegistry.CABLE);
    public static final BlockEntityType<StorageCableBlockEntity> STORAGE_CABLE = create(StorageCableBlockEntity::new, BlockRegistry.STORAGE_CABLE);

    public static void register() {
        register(BlockRegistry.MASTER, MASTER);
        register(BlockRegistry.CRAFTING_TERMINAL, CRAFTING_TERMINAL);
        register(BlockRegistry.PROCESSING_TERMINAL, PROCESSING_TERMINAL);
        register(BlockRegistry.CABLE, CABLE);
        register(BlockRegistry.STORAGE_CABLE, STORAGE_CABLE);
    }

    protected static void register(ModBlock block, BlockEntityType<? extends BlockEntity> blockEntityType) {
        Registry.register(Registry.BLOCK_ENTITY_TYPE, block.ID, blockEntityType);
    }

    private static <T extends BlockEntity> BlockEntityType<T> create(Supplier<T> supplier, Block block) {
        return BlockEntityType.Builder.create(supplier, block).build(null);
    }

}

File: src/main/kotlin/badasintended/slotlink/block/entity/BlockEntityTypeRegistry.kt
package badasintended.slotlink.block.entity

import badasintended.slotlink.block.BlockRegistry
import badasintended.slotlink.block.ModBlock
import net.minecraft.block.Block
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.util.registry.Registry
import java.util.function.Supplier

object BlockEntityTypeRegistry {

    val MASTER = c(BlockRegistry.MASTER) { MasterBlockEntity() }
    val REQUEST = c(BlockRegistry.REQUEST) { RequestBlockEntity() }

    val CABLE = c(BlockRegistry.CABLE) { CableBlockEntity() }
    val LINK_CABLE = c(BlockRegistry.LINK_CABLE) { LinkCableBlockEntity() }

    fun init() {
        r(BlockRegistry.MASTER, MASTER)
        r(BlockRegistry.REQUEST, REQUEST)

        r(BlockRegistry.CABLE, CABLE)
        r(BlockRegistry.LINK_CABLE, LINK_CABLE)
    }

    private fun r(modBlock: ModBlock, blockEntityType: BlockEntityType<out BlockEntity>) {
        Registry.register(Registry.BLOCK_ENTITY_TYPE, modBlock.id, blockEntityType)
    }

    private fun <T : BlockEntity> c(block: Block, function: () -> T): BlockEntityType<T> {
        return BlockEntityType.Builder.create(Supplier(function), block).build(null)
    }

}
